examples/kprobe/main.go

```diff
@@ -0,0 +1,106 @@
+// This program demonstrates how to attach an eBPF program to a kprobe.
+// The program will be attached to the __x64_sys_execve syscall and print out
+// the number of times it has been called every second.
+package main
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"os/signal"
+	"runtime"
+	"syscall"
+	"time"
+	"unsafe"
+
+	goperf "github.com/elastic/go-perf"
+	"golang.org/x/sys/unix"
+)
+
+//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -cc clang-11 KProbeExample ./bpf/kprobe_example.c -- -I../headers
+
+const mapKey uint32 = 0
+
+func main() {
+	stopper := make(chan os.Signal, 1)
+	signal.Notify(stopper, os.Interrupt, syscall.SIGTERM)
+
+	// Increase rlimit so the eBPF map and program can be loaded.
+	if err := unix.Setrlimit(unix.RLIMIT_MEMLOCK, &unix.Rlimit{
+		Cur: unix.RLIM_INFINITY,
+		Max: unix.RLIM_INFINITY,
+	}); err != nil {
+		log.Fatalf("failed to set temporary rlimit: %v", err)
+	}
+
+	// Load Program and Map
+	specs, err := NewKProbeExampleSpecs()
+	if err != nil {
+		log.Fatalf("error while loading specs: %v", err)
+	}
+	objs, err := specs.Load(nil)
+	if err != nil {
+		log.Fatalf("error while loading objects: %v", err)
+	}
+
+	// Create and attach __x64_sys_execve kprobe
+	efd, err := openKProbe("__x64_sys_execve", uint32(objs.ProgramKprobeExecve.FD()))
+	if err != nil {
+		log.Fatalf("create and attach KProbe: %v", err)
+
+	}
+	defer unix.Close(efd)
+
+	ticker := time.NewTicker(1 * time.Second)
+	for {
+		select {
+		case <-ticker.C:
+			var value uint64
+			if err := objs.MapKprobeMap.Lookup(mapKey, &value); err != nil {
+				log.Fatalf("error while reading map: %v", err)
+			}
+			log.Printf("__x64_sys_execve called %d times\n", value)
+		case <-stopper:
+			return
+		}
+	}
+}
+
+func openKProbe(syscall string, fd uint32) (int, error) {
+	et, err := goperf.LookupEventType("kprobe")
+	if err != nil {
+		return 0, fmt.Errorf("read PMU type: %v", err)
+	}
+
+	config1ptr := uintptr(newStringPointer(syscall))
+	ev, err := goperf.Open(&goperf.Attr{Type: et, Config1: uint64(config1ptr)}, goperf.AllThreads, 0, nil)
+	if err != nil {
+		return 0, fmt.Errorf("perf event open: %v", err)
+	}
+	efd, err := ev.FD()
+	if err != nil {
+		return 0, fmt.Errorf("get perf event fd: %v", err)
+	}
+
+	// Ensure config1ptr is not finalized until goperf.Open returns.
+	runtime.KeepAlive(config1ptr)
```

Comment:
This doesn't work with `uintptr`, you have to use an `unsafe.Pointer`. Otherwise the runtime doesn't know that this integer value is a location in memory that has to be kept alive.